tools/pubsys/src/aws/validate_ami/ami.rs

```diff
@@ -0,0 +1,195 @@
+        .collect()
+        .await
+}
+
+/// Fetches all images in a single region
+pub(crate) async fn describe_images_in_region(
+    region: &Region,
+    client: &Ec2Client,
+    image_ids: Vec<String>,
+    expected_image_public: &HashMap<String, bool>,
+) -> Result<HashMap<String, ImageDef>> {
+    info!("Retrieving images in {}", region.to_string());
+    let mut images = HashMap::new();
+
+    // Send the request
+    let mut get_future = client
+        .describe_images()
+        .include_deprecated(true)
+        .set_image_ids(Some(image_ids))
+        .into_paginator()
+        .send();
+
+    // Iterate over the retrieved images
+    while let Some(page) = get_future.next().await {
+        let retrieved_images = page
+            .context(error::DescribeImagesSnafu {
+                region: region.to_string(),
+            })?
+            .images()
+            .unwrap_or_default()
+            .to_owned();
+        for image in retrieved_images {
+            // Insert a new key-value pair into the map, with the key containing image id
+            // and the value containing the ImageDef object created from the image
+            let image_id = image
+                .image_id()
+                .ok_or(error::Error::MissingField {
+                    missing: "image_id".to_string(),
+                })?
+                .to_string();
+            let expected_public = expected_image_public.get(&image_id).ok_or(
+                error::Error::MissingExpectedPublic {
+                    missing: image_id.clone(),
+                },
+            )?;
+            // If the image is not expected to be public, retrieve the launch permissions
+            trace!(
+                "Retrieving launch permissions for {} in {}",
+                image_id,
+                region.as_ref()
+            );
+            let mut image_def = ImageDef::from(image.to_owned());
+            if !*expected_public {
+                image_def.launch_permissions = Some(
+                    get_launch_permissions(client, region.as_ref(), &image_id)
+                        .await
+                        .context(error::GetLaunchPermissionsSnafu {
+                            region: region.as_ref(),
+                            image_id: image_id.clone(),
+                        })?,
+                );
+            }
+            images.insert(image_id, image_def);
+        }
+    }
+
+    info!("Images in {} have been retrieved", region.to_string());
+    Ok(images)
+}
+
+pub(crate) mod error {
+    use aws_sdk_ec2::error::DescribeImagesError;
+    use aws_sdk_ssm::types::SdkError;
+    use snafu::Snafu;
+
+    #[derive(Debug, Snafu)]
+    #[snafu(visibility(pub(super)))]
+    #[allow(clippy::large_enum_variant)]
+    pub(crate) enum Error {
+        #[snafu(display("Failed to describe images in {}: {}", region, source))]
+        DescribeImages {
+            region: String,
+            source: SdkError<DescribeImagesError>,
+        },
```

Comment:
Where-ever we're using a `SdkError` as source we need to do this to get the full error message:

```suggestion
use aws_smithy_types::error::display::DisplayErrorContext;

        #[snafu(display("Failed to describe images in {}: {}", region, DisplayErrorContext(source)))]
        DescribeImages {
            region: String,
            source: SdkError<DescribeImagesError>,
        },
```

See https://docs.rs/aws-smithy-client/latest/aws_smithy_client/enum.SdkError.html
> When logging an error from the SDK, it is recommended that you either wrap the error in [DisplayErrorContext](https://docs.rs/aws-smithy-types/0.55.1/x86_64-unknown-linux-gnu/aws_smithy_types/error/display/struct.DisplayErrorContext.html), use another error reporter library that visits the error’s cause/source chain, or call [Error::source](https://doc.rust-lang.org/nightly/core/error/trait.Error.html#method.source) for more details about the underlying cause.

**Note: This needs to be a tree wide change, but might as well start here.** 

Please create an issue for the other places if you're willing.